sdk/storage/Azure.Storage.Internal.Avro/src/AvroParser.cs

```diff
@@ -0,0 +1,472 @@
+            return BitConverter.ToDouble(bytes, 0);
+        }
+
+        /// <summary>
+        /// Reads a fixed number of bytes from the stream.
+        /// </summary>
+        public static async Task<byte[]> ReadBytesAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            // Note that byte array length is actually defined as a long in the Avro spec.
+            // This is fine for now, but may need to be changed in the future.
+            int size = await ReadIntAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            return await ReadFixedBytesAsync(stream, size, async, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Reads a string from the stream.
+        /// </summary>
+        public static async Task<string> ReadStringAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] bytes = await ReadBytesAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            return Encoding.UTF8.GetString(bytes);
+        }
+
+        /// <summary>
+        /// Reads a KeyValuePair from the stream.
+        /// Used in ReadMapAsync().
+        /// </summary>
+        private static async Task<KeyValuePair<string, T>> ReadMapPairAsync<T>(
+            Stream stream,
+            Func<Stream, bool, CancellationToken, Task<T>> parseItemAsync,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            string key = await ReadStringAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            #pragma warning disable AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+            T value = await parseItemAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            #pragma warning restore AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+            return new KeyValuePair<string, T>(key, value);
+        }
+
+        /// <summary>
+        /// Reads a map from the stream.
+        /// </summary>
+        public static async Task<Dictionary<string, T>> ReadMapAsync<T>(
+            Stream stream,
+            Func<Stream, bool, CancellationToken, Task<T>> parseItemAsync,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            #pragma warning disable AZC0110 // DO NOT use await keyword in possibly synchronous scope.
+            #pragma warning disable AZC0108 // Incorrect 'async' parameter value.
+            Func<Stream, bool, CancellationToken, Task<KeyValuePair<string, T>>> parsePair =
+                async (s, a, c) => await ReadMapPairAsync(s, parseItemAsync, a, c).ConfigureAwait(false);
+            #pragma warning restore AZC0108 // Incorrect 'async' parameter value.
+            #pragma warning restore AZC0110 // DO NOT use await keyword in possibly synchronous scope.
```

Comment:
Nit - If we change `a` and `c` to `async` and `cancellationToken` it might make these errors go away.

Reply:
Fixed.